selfhost/typechecker.jakt

```diff
@@ -1460,6 +1548,334 @@ struct Typechecker {
+                match rhs_type {
+                    RawPtr(rhs_rawptr_type_id) => {
+                        .check_types_for_compat(
+                            lhs_type_id: lhs_rawptr_type_id
+                            rhs_type_id: rhs_rawptr_type_id
+                            generic_inferences
+                            span
+                        )
+                    }
+                    else => {
+                        if not rhs_type_id.equals(lhs_type_id) {
+                            .error(
+                                format("Type mismatch: expected `{}` but got `{}`", .type_name(lhs_type_id), .type_name(rhs_type_id))
+                                span
+                            )
+                        }
+                    }
+                }
+            }
+            else => {
+                if not rhs_type_id.equals(lhs_type_id) {
+                    .error(
+                        format("Type mismatch: expected `{}` but got `{}`", .type_name(lhs_type_id), .type_name(rhs_type_id))
+                        span
+                    )
+                }
+            }
+        }
+
+        return ret
+    }
+
+    // FIXME: Use [TypeId: TypeID] without TypeId.to_string()/TypeId::from_string() workaround
+    function substitute_typevars_in_type(mut this, type_id: TypeId , generic_inferences: [String: String]) throws -> TypeId {
+        mut result = .substitute_typevars_in_type_helper(type_id, generic_inferences)
+
+        loop {
+            let fixed_point = .substitute_typevars_in_type_helper(type_id, generic_inferences)
+
+            if fixed_point.equals(result) {
+                break
+            } else {
+                result = fixed_point
+            }
+        }
+        return result
+    }
+
+    // FIXME: Use [TypeId: TypeID] without TypeId.to_string()/TypeId::from_string() workaround
+    function substitute_typevars_in_type_helper(mut this, type_id: TypeId , generic_inferences: [String: String]) throws -> TypeId {
+        let type_ = .get_type(type_id)
+
+        match type_ {
+            TypeVariable() => {
+                let replacment_type_id_string = generic_inferences.get(type_id.to_string())
+                if replacment_type_id_string.has_value() {
+                    return TypeId::from_string(replacment_type_id_string.value())
+                }
+            }
+            GenericInstance(id, args) => {
+                mut new_args = args
+                for arg in new_args.iterator() {
+                    arg = .substitute_typevars_in_type(type_id: arg, generic_inferences)
+                }
+
+                return .find_or_add_type_id(Type::GenericInstance(id, args: new_args))
+            }
+            GenericEnumInstance(id, args) => {
+                mut new_args = args
+                for arg in new_args.iterator() {
+                    arg = .substitute_typevars_in_type(type_id: arg, generic_inferences)
```

Comment:
This does not actually change `args`. Try running this test:
```
struct TypeId {
    id: u64
}

function main() {
    mut args = [TypeId(id: 0); 2]

    for arg in args.iterator() {
        arg.id = 2
    }

    println("{}", args)
}
```
You'll see that `args` is still the same as before even though we changed `arg`. You need to index into `new_args` and assign to that. So something like `new_args[index] = .substitute_typevars_in_type(...)`.
The same holds applies to the `Struct` and `Enum` cases below :^)

Reply:
good catch, as we talked this would change `args` by reference, so I made new emtpy new_args with ensure capacity, and push the results to the new one.